### PR TITLE
[Awaiting Review] Fix Metadata - Integration tests passing on MOL

### PR DESCRIPTION
**Pull request by xtalax**
I had made a mistake on https://github.com/SciML/ModelingToolkit.jl/pull/1768, neglecting to add `nothing` to pad the arguments of the constructors to the systems so that metadata was in the right place. 
This fixes that issue.
This has been tested locally in MOL, and works there.

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Can you add tests?

**Comment by xtalax** [comment]
Tested

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by YingboMa** [comment]
CI and format checker fails.

File: src/ModelingToolkit.jl [hunk 1]
```diff
@@ -59,6 +59,7 @@ import Symbolics: rename, get_variables!, _solve, hessian_sparsity,
                   substituter, scalarize, getparent
 
 import DiffEqBase: @add_kwonly
+import SciMLBase: StandardODEProblem
```
Comment (YingboMa) [inline src/ModelingToolkit.jl:62]: Where is this used?

**Review by YingboMa — CHANGES_REQUESTED** [review]
Won't it be easier if you just reorder the metadata field?

**Comment by xtalax** [comment]
> Won't it be easier if you just reorder the metadata field?

Leads to this error when reordered
```
  type TearingState has no field deps
  Stacktrace:
    [1] getproperty(x::TearingState{NonlinearSystem}, f::Symbol)
      @ Base ./Base.jl:33
    [2] empty_substitutions(sys::NonlinearSystem)
      @ ModelingToolkit ~/.julia/dev/ModelingToolkit.jl/src/utils.jl:541
    [3] get_substitutions_and_solved_states(sys::NonlinearSystem; no_postprocess::Bool)
      @ ModelingToolkit ~/.julia/dev/ModelingToolkit.jl/src/utils.jl:545
    [4] get_substitutions_and_solved_states
      @ ~/.julia/dev/ModelingToolkit.jl/src/utils.jl:545 [inlined]
    [5] generate_function(sys::NonlinearSystem, dvs::Vector{Term{Real, Nothing}}, ps::Vector{Any}; kwargs::Base.Iterators.Pairs{Symbol, Any, NTuple{5, Symbol}, NamedTuple{(:expression, :checkbounds, :linenumbers, :parallel, :check_length), Tuple{DataType, Bool, Bool, Nothing, Bool}}})
      @ ModelingToolkit ~/.julia/dev/ModelingToolkit.jl/src/systems/nonlinear/nonlinearsystem.jl:174
    [6] (NonlinearFunction{true, recompile, F, TMM, Ta, Tt, TJ, JVP, VJP, JP, SP, TW, TWt, TPJ, S, O, TCV, SYS} where {recompile, F, TMM, Ta, Tt, TJ, JVP, VJP, JP, SP, TW, TWt, TPJ, S, O, TCV, SYS})(sys::NonlinearSystem, dvs::Vector{Term{Real, Nothing}}, ps::Vector{Any}, u0::Vector{Float64}; version::Nothing, jac::Bool, eval_expression::Bool, sparse::Bool, simplify::Bool, kwargs::Base.Iterators.Pairs{Symbol, Union{Nothing, Bool}, NTuple{4, Symbol}, NamedTuple{(:checkbounds, :linenumbers, :parallel, :check_length), Tuple{Bool, Bool, Nothing, Bool}}})
      @ ModelingToolkit ~/.julia/dev/ModelingToolkit.jl/src/systems/nonlinear/nonlinearsystem.jl:215
    [7] process_NonlinearProblem(constructor::Type, sys::NonlinearSystem, u0map::Vector{Float64}, parammap::SciMLBase.NullParameters; version::Nothing, jac::Bool, checkbounds::Bool, sparse::Bool, simplify::Bool, linenumbers::Bool, parallel::Nothing, eval_expression::Bool, use_union::Bool, kwargs::Base.Iterators.Pairs{Symbol, Bool, Tuple{Symbol}, NamedTuple{(:check_length,), Tuple{Bool}}})
      @ ModelingToolkit ~/.julia/dev/ModelingToolkit.jl/src/systems/nonlinear/nonlinearsystem.jl:319
    [8] (NonlinearProblem{true, isinplace, P, F, K} where {isinplace, P, F, K})(sys::NonlinearSystem, u0map::Vector{Float64}, parammap::SciMLBase.NullParameters; check_length::Bool, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
      @ ModelingToolkit ~/.julia/dev/ModelingToolkit.jl/src/systems/nonlinear/nonlinearsystem.jl:345
    [9] NonlinearProblem (repeats 2 times)
      @ ~/.julia/dev/ModelingToolkit.jl/src/systems/nonlinear/nonlinearsystem.jl:345 [inlined]
   [10] #NonlinearProblem#607
      @ ~/.julia/dev/ModelingToolkit.jl/src/systems/nonlinear/nonlinearsystem.jl:339 [inlined]
   [11] NonlinearProblem(sys::NonlinearSystem, args::Vector{Float64})
      @ ModelingToolkit ~/.julia/dev/ModelingToolkit.jl/src/systems/nonlinear/nonlinearsystem.jl:339
...
```

**Comment by YingboMa** [comment]
You need to reorder the constructor arguments as well.

**Comment by xtalax** [comment]
I did, of course. Changing the internal constructor arguments and the call to it in the external constructor, leads to this error. I think this shows that it will not be easier to just reorder the constructor.

**Comment by xtalax** [comment]
IMPORTANT: this depends on https://github.com/SciML/SciMLBase.jl/pull/246

File: src/ModelingToolkit.jl (hunk 1 above)
Comment (xtalax) [reply src/ModelingToolkit.jl]: as the default for problem_type when metadata is `nothing`

**Comment by xtalax** [comment]
@YingboMa I think it's definitely easier to leave the constructor as is, it seems that the inner constructor is called in multiple places and your requested change would require tracking down each of those calls - I don't know where these are, errors occur later than this/these call/calls so this is hard for me to do without complete knowledge of the package.

File: src/systems/nonlinear/nonlinearsystem.jl [hunk 2]
```diff
@@ -344,7 +344,12 @@ function DiffEqBase.NonlinearProblem{iip}(sys::NonlinearSystem, u0map,
                                           check_length = true, kwargs...) where {iip}
     f, u0, p = process_NonlinearProblem(NonlinearFunction{iip}, sys, u0map, parammap;
                                         check_length, kwargs...)
-    NonlinearProblem{iip}(f, u0, p; kwargs...)
+    pt = if get_metadata(sys) === nothing
```
Comment (YingboMa) [inline src/systems/nonlinear/nonlinearsystem.jl:347]: `something(get_metadata(sys), StandardNonlinearProblem())`?

Comment (xtalax) [reply src/systems/nonlinear/nonlinearsystem.jl:347]: I didn't know about something, thanks.

**Comment by ChrisRackauckas** [comment]
Tests fail?

File: src/ModelingToolkit.jl [hunk 3]
```diff
@@ -183,7 +184,7 @@ export Equation, ConstrainedEquation
 export Term, Sym
 export SymScope, LocalScope, ParentScope, GlobalScope
 export independent_variables, independent_variable, states, parameters, equations, controls,
-       observed, structure, full_equations
+       observed, structure, full_equations, get_metadata
```
Comment (YingboMa) [inline src/ModelingToolkit.jl:187]: No other `get_` functions are exported.

File: test/optimizationsystem.jl [hunk 4]
```diff
@@ -126,6 +126,18 @@ OBS2 = OBS
...
+@testset "Test Metadata in constructor" begin
+    @variables x
+    o1 = (x - 1)^2
+    c1 = [
+        x ~ 1,
+    ]
+    testdict = Dict(["test" => 1])
+    sys1 = OptimizationSystem(o1, [x], [], name = :sys1, constraints = c1,
+                              metadata = testdict)
+    @test sys1.metadata == testdict
```
Comment (YingboMa) [inline test/optimizationsystem.jl:138]: Don't use `metadata`.

File: test/nonlinearsystem.jl [hunk 5]
```diff
@@ -202,3 +202,13 @@ let
...
 end
+
+@testset "Test Metadata in constructor" begin
+    @variables x(t)
+    @parameters a
+    eqs = [0 ~ a * x]
+
+    testdict = Dict([:test => 1])
+    @named sys = NonlinearSystem(eqs, [x], [a], name = :test, metadata = testdict)
+    @test sys.metadata == testdict
```
Comment (YingboMa) [inline test/nonlinearsystem.jl:213]: same here.

**Review by YingboMa — CHANGES_REQUESTED** [review]
What about `metadata` for other systems? Also, you should move the metadata field before the internal fields, so you won't need `nothing, nothing`.

**Pull request closed without merging**